drivers/watchdog/wdt_sam4l.c

```diff
@@ -0,0 +1,341 @@
+		return -ENOMEM;
+	}
+
+	if (config->flags & WDT_FLAG_RESET_CPU_CORE) {
+		LOG_ERR("The SAM4L watchdog do not support reset CPU core");
+		return -ENOTSUP;
+	}
+
+	if (wdt_sam4l_convert_timeout(cfg, config, &tban, &psel)) {
+		return -EINVAL;
+	}
+
+	reg = wdt->CTRL;
+
+	if (reg & WDT_CTRL_EN) {
+		LOG_ERR("Watchdog is running and can not be changed");
+		return -EPERM;
+	}
+
+	if (reg & WDT_CTRL_SFV) {
+		LOG_ERR("Watchdog is locked and can not be changed");
+		return -EPERM;
+	}
+
+	data->cb = config->callback;
+	data->flags = config->flags;
+	reg = (cfg->is_osc32k ? WDT_CTRL_CSSEL : 0)
+	    | (config->callback ? WDT_CTRL_IM : 0)
+	    | (config->window.min ? WDT_CTRL_MODE : 0)
+	    | (config->window.min ? WDT_CTRL_TBAN(tban) : 0)
+	    | WDT_CTRL_PSEL(psel);
+	wdt_sam4l_set_ctrl(cfg, reg);
+
+	wdt_sam4l_set_ctrl(cfg, wdt->CTRL | WDT_CTRL_CEN);
+	while (!(wdt->CTRL & WDT_CTRL_CEN)) {
+	}
+
+	return 0;
+}
+
+static int wdt_sam4l_feed(const struct device *dev, int channel_id)
+{
+	const struct wdt_sam4l_dev_cfg *cfg = dev->config;
+	Wdt *const wdt = cfg->regs;
+
+	ARG_UNUSED(channel_id);
+
+	while ((wdt->SR & WDT_SR_CLEARED) != WDT_SR_CLEARED) {
+	}
+	wdt->CLR = WDT_CLR_WDTCLR
+		 | WDT_CLR_KEY(WDT_FIRST_KEY);
+	wdt->CLR = WDT_CLR_WDTCLR
+		 | WDT_CLR_KEY(WDT_SECOND_KEY);
+
+	return 0;
+}
+
+/* If the callback blocks or ISR do not clear flags the system will trigger a
+ * CPU reset on the next watchdog timeout, see 20.5.3 Interrupt Mode
+ */
+static void wdt_sam4l_isr(const struct device *dev)
+{
+	const struct wdt_sam4l_dev_cfg *cfg = dev->config;
+	struct wdt_sam4l_dev_data *const data = dev->data;
+	Wdt *const wdt = cfg->regs;
+
+	wdt->ICR = WDT_ICR_WINT;
+	data->cb(dev, 0);
+}
+
+static DEVICE_API(wdt, wdt_sam4l_driver_api) = {
+	.setup = wdt_sam4l_setup,
+	.disable = wdt_sam4l_disable,
+	.install_timeout = wdt_sam4l_install_timeout,
+	.feed = wdt_sam4l_feed,
+};
+
+static int wdt_sam4l_init(const struct device *dev)
+{
+	const struct wdt_sam4l_dev_cfg *const cfg = dev->config;
+	Wdt *const wdt = cfg->regs;
+
+	/* Enable WDT clock in PMC */
+	(void)clock_control_on(SAM_DT_PMC_CONTROLLER,
+			       (clock_control_subsys_t)&cfg->clock_cfg);
+
+	if (IS_ENABLED(CONFIG_WDT_DISABLE_AT_BOOT)) {
+		wdt_sam4l_disable(dev);
+		return 0;
+	}
+
+	wdt->IDR = WDT_IDR_MASK;
+	cfg->irq_cfg_func(dev);
+	wdt->IER = WDT_IER_WINT;
+
+	return 0;
+}
+
+#define WDT_SAM4L_INIT(n)						\
+	static void wdt##n##_sam4l_irq_cfg(const struct device *dev)	\
+	{								\
+		IRQ_CONNECT(DT_INST_IRQN(n),				\
+			    DT_INST_IRQ(n, priority),			\
+			    wdt_sam4l_isr,				\
+			    DEVICE_DT_INST_GET(n), 0);			\
+		irq_enable(DT_INST_IRQN(n));				\
+	}								\
+									\
+	static const struct wdt_sam4l_dev_cfg wdt##n##_sam4l_cfg = {	\
+		.regs = (Wdt *)DT_INST_REG_ADDR(n),			\
+		.clock_cfg = SAM_DT_INST_CLOCK_PMC_CFG(n),		\
+		.irq_cfg_func = wdt##n##_sam4l_irq_cfg,			\
+		.lock = DT_INST_PROP(n, lock_mode),			\
+		.is_osc32k = false,					\
```

Comment:
No way to configure this?

Reply:
If you referring on `is_osc32k` the OSC32K is not enabled yet so it is not possible to configure it yet : )

Reply:
I mean there is no way to enable it, as it's located in config (`const`) so it either needs to move to the data or have a device tree binding?

Reply:
In the future I want to pick from devicetree but there is no point to enable at moment because the osc32k is not yet available. If this is too annoying I can drop it.

Reply:
I'd suggest adding the binding nonetheless, doesn't harm and might be useful for out-of-tree boards?

Reply:
I agreed with you, done!